comm/src/main/scala/coop/rchain/comm/connect/Connect.scala

```diff
@@ -120,25 +116,36 @@ object Connect {
   private def handleProtocolHandshake[
       F[_]: Monad: Time: TransportLayer: NodeDiscovery: Log: ErrorHandler](
       peer: PeerNode,
-      maybePh: Option[ProtocolHandshake]): F[CommunicationResponse] =
+      maybePh: Option[ProtocolHandshake],
+      defaultTimeout: FiniteDuration
+  ): F[CommunicationResponse] =
     for {
-      local    <- TransportLayer[F].local
-      _        <- getOrError[F, ProtocolHandshake](maybePh, parseError("ProtocolHandshake"))
-      phr      = protocolHandshakeResponse(local)
-      addedErr <- NodeDiscovery[F].addNode(peer)
-      commResponse <- addedErr.fold(
+      local  <- TransportLayer[F].local
+      _      <- getOrError[F, ProtocolHandshake](maybePh, parseError("ProtocolHandshake"))
+      hbrErr <- TransportLayer[F].roundTrip(peer, heartbeat(local), defaultTimeout)
+      commResponse <- hbrErr.fold(
                        error =>
-                         Log[F].error(s"Could not add $peer, reason: $error").as(notHandled(error)),
+                         Log[F].warn(s"Not adding. Could receive Pong message back from $peer, reason: $error").as(notHandled(error)),
                        _ =>
-                         Log[F]
-                           .info(s"Responded to protocol handshake request from $peer")
-                           .as(handledWithMessage(phr))
+                         NodeDiscovery[F].addNode(peer) *>
+                           Log[F]
+                             .info(s"Responded to protocol handshake request from $peer")
+                             .as(handledWithMessage(protocolHandshakeResponse(local)))
                      )
     } yield commResponse
 
+  private def handleHeartbeat[F[_]: Monad: TransportLayer: ErrorHandler](
+      peer: PeerNode,
+      maybeHeartbeat: Option[Heartbeat]): F[CommunicationResponse] =
```

Comment:
I don't get it. Why do we get _maybe_ messages? Can't we bounce invalid messages earlier? IMO `Option` method parameters are a code smell.

Reply:
Check how that method is being used. This is the consequence of using `proto.Any` in `routing.proto`. This will eventually go away once that `proto.Any` is gone.